src/shims/unix/fd.rs

```diff
@@ -618,22 +624,67 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         let bytes = this.read_bytes_ptr_strip_provenance(buf, Size::from_bytes(count))?.to_owned();
         // We temporarily dup the FD to be able to retain mutable access to `this`.
         let Some(fd) = this.machine.fds.get(fd_num) else {
-            return Ok(Scalar::from_target_isize(this.fd_not_found()?, this));
+            let res: i32 = this.fd_not_found()?;
+            this.write_int(res, dest)?;
+            return Ok(());
         };
 
-        let result = match offset {
-            None => fd.write(&fd, communicate, &bytes, this),
+        match offset {
+            None => fd.write(&fd, communicate, &bytes, dest, this)?,
             Some(offset) => {
                 let Ok(offset) = u64::try_from(offset) else {
                     let einval = this.eval_libc("EINVAL");
                     this.set_last_error(einval)?;
-                    return Ok(Scalar::from_target_isize(-1, this));
+                    this.write_int(-1, dest)?;
+                    return Ok(());
                 };
-                fd.pwrite(communicate, &bytes, offset, this)
+                fd.pwrite(communicate, &bytes, offset, dest, this)?
             }
         };
+        Ok(())
+    }
+
+    /// This function either writes to the user supplied buffer and to dest place, or return error.
+    fn read_byte_helper(
+        &mut self,
+        buf: Pointer,
+        bytes: Vec<u8>,
+        result: io::Result<usize>,
+        dest: &MPlaceTy<'tcx>,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        // `File::read` never returns a value larger than `count`, so this cannot fail.
+        match result.map(|c| i64::try_from(c).unwrap()) {
```

Comment:
```suggestion
        match result {
```

This conversion isn't necessary I think?

Reply:
Yup, we can use ``usize`` directly.